content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+        "ins_replace", 
+        Type::getDoubleTy(Ctx),
+        Type::getDoubleTy(Ctx), Type::getDoubleTy(Ctx), Type::getInt32Ty(Ctx)
+        );
+    // function call name, return type and argument types have to be known
+
+    for (auto& B : F) {
+        queue <Instruction*> workList; // collects replaced instructions within iterator
+        for (auto& I : B) {
+            /* Add runtime instruction selection here */
+            
+            /* End of selection algorithm */
+                
+            /* find the place to enter the runtime call */
+            IRBuilder<> builder(I);
+            builder.SetInsertPoint(&B, ++builder.GetInsertPoint());
+            
+            
+            /* insert runtime call, collect return value */
+            Constant *i32_config = ConstantInt::get(Type::getInt32Ty(Ctx), /* configuration */, true);
+            Value* args[] = {I.getOperand(0), I.getOperand(1), i32_config};
+            Value* ret =  builder.CreateCall(rtFunc, args);
+            
+            /* forward return to users */
+            for (auto& U : I.uses()) {
+                User* user = U.getUser();
+                user->setOperand(U.getOperandNo(), ret);
+            }
+            
+            /* mark instruction for deletion */
+            workList.push(&I);
+            I.dropAllReferences();
+        }
+        
+        /* delete replaced instructions */
+        while (!workList.empty()) {
+            workList.front()->eraseFromParent();
+            workList.pop();
+        }
+    }
+```
+Key limitations to note, runtime library has to be co-designed with the function declarations 
+baked into the llvm pass and of course, the instruction selection algorithm can get very complex
+depending on how interesting the task is.
+
+I wish to direct your attention also to the granularity of the replacements, the runtime functions
+are essentially overloading the llvm instructions here. This granularity is not particulary useful for
+loosely-coupled accelerators, as communication between the accelerator and the CPU is at the highest.
+This is reflected in the performance section.
+
+## Replacing functions
+In contrast to the instruction level, functions are a substantially coarser-grained abstraction 
+for an accelerator to replace. Even a loosely-coupled accelerator can tradeoff the communication latencies
+by efficient computation. 
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+        "func_replace", 
+        Type::getVoidTy(Ctx),
+        Type::getDoublePtrTy(Ctx), Type::getDoublePtrTy(Ctx), Type::getDoublePtrTy(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    /* Add runtime function selection here */
```

Comment:
As above, it seems important to explain what code you're assuming will be provided.